#include "mainwindow.h"
#include "ui_mainwindow.h"

#include <QDebug>

#include <QFileDialog>

#include <QMessageBox>

#include<QDialog>
#include<QColorDialog>
#include<QFontDialog>

#include  <QPainter>
#include <QImage>
#include <QPixmap>
#include <QBitmap>
#include <QPointF>
#include <QColor>

#include <QMenu>
#include <QMenuBar>
#include <QAction>

#include <QGraphicsScene>
#include <QGraphicsView>

#include <QTextCodec>

#include <QtMath>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);
    mCheckList<<"CRC-16"<<QString::fromLocal8Bit("和校验-16");
    ui->comboBox_check->addItems(mCheckList);
    ui->comboBox_check->setCurrentIndex(1);
}

MainWindow::~MainWindow()
{
    delete ui;
}

char fileSrc[FILE_SIZE_MAX];
QString outFilePath;
quint32 fileLength;
void MainWindow::on_pushButton_chooseFile_clicked()
{
    fileLength = fun_getFileSource(fileSrc);
    if(fileLength == 0)
    {
        ui->pushButton_save->setEnabled(false);
    }
    else
    {
        ui->pushButton_save->setEnabled(true);
    }

}

unsigned int  MainWindow::fun_getFileSource(char *dst)
{
    QString filePath = QFileDialog::getOpenFileName(this, QString::fromLocal8Bit("请选择数据文件"), NULL, "program file(*.bin *.BIN)");
    if (filePath.isEmpty())
    {
        QMessageBox::warning(this, QString::fromLocal8Bit("提示"), QString::fromLocal8Bit("未选择数据文件"), QString::fromLocal8Bit("确定"));
        return 0;
    }
    else
    {
        QFile openFile(filePath);
        if(!openFile.open(QIODevice::ReadOnly))
        {
            openFile.close();
            QMessageBox::warning(this, QString::fromLocal8Bit("提示"), QString::fromLocal8Bit("文件打开失败"), QString::fromLocal8Bit("确定"));
            filePath.clear();
            return 0;
        }

        unsigned int thisFileLength = openFile.size();
        if(thisFileLength > FILE_SIZE_MAX)
        {
            openFile.close();
            QMessageBox::warning(this, QString::fromLocal8Bit("提示"), QString::fromLocal8Bit("文件大小超过限制4MB"), QString::fromLocal8Bit("确定"));
            filePath.clear();
            return 0;
        }

        QApplication::setOverrideCursor(Qt::WaitCursor);

        ui->listWidget_inPath->clear();//显示路径
        ui->listWidget_inPath->addItem(filePath);

        openFile.seek(0);
        openFile.read(dst,thisFileLength);

        QApplication::restoreOverrideCursor();

        openFile.close();

        QString outFileName = openFile.fileName();
        outFileName.remove(outFileName.size()-4,4);
        if(ui->comboBox_check->currentIndex() == 0)
        {
            outFileName+= "CRC16.bin";
        }
        else if(ui->comboBox_check->currentIndex() == 1)
        {
            outFileName+= "SUM16.bin";
        }

        outFilePath  = filePath;
        outFilePath.remove(filePath.length() - openFile.fileName().length(),openFile.fileName().length());
        outFilePath +=  outFileName;

        return thisFileLength;
    }
}

void MainWindow::on_pushButton_save_clicked()
{
    if(ui->comboBox_check->currentIndex() == 0)
    {
        unsigned short resCRC16;
        quint8 resCRCH = 0xFF;  /* CRC  的高字节初始化  */
        quint8 resCRCL = 0xFF;  /* CRC  的低字节初始化  */
        resCRC16 = GetCRC16(fileSrc,fileLength);
        resCRCH = (resCRC16>>8)&0xFF;
        resCRCL = (resCRC16)&0xFF;
        *(fileSrc + fileLength) = resCRCH;
        *(fileSrc + fileLength + 1) = resCRCL;
    }
    else if(ui->comboBox_check->currentIndex() == 1)
    {
        unsigned short resSUM16;
        quint8 resSUMH = 0xFF;  /* CRC  的高字节初始化  */
        quint8 resSUML = 0xFF;  /* CRC  的低字节初始化  */
        resSUM16 = GetSUM16(fileSrc,fileLength);
        resSUMH = (resSUM16>>8)&0xFF;
        resSUML = (resSUM16)&0xFF;
        *(fileSrc + fileLength) = resSUMH;
        *(fileSrc + fileLength + 1) = resSUML;
    }


    fun_saveToFile(outFilePath,(char *)fileSrc,fileLength + 2);

}

void MainWindow::fun_saveToFile(QString filePath, char * buf, quint64 len)
{
    QFile saveFile(filePath);
    saveFile.remove();
    if (!saveFile.open(QIODevice::NewOnly|QIODevice::ReadWrite|QIODevice::Truncate))
    {
        QMessageBox::warning(this, QString::fromLocal8Bit("提示"), QString::fromLocal8Bit("文件保存失败"), QString::fromLocal8Bit("确定"));
        saveFile.close();
        return;
    }
    QApplication::setOverrideCursor(Qt::WaitCursor);
    saveFile.seek(0);
    saveFile.write(buf, len);
    saveFile.seek(0);

    QApplication::restoreOverrideCursor();
    saveFile.close();
    QMessageBox::warning(this, QString::fromLocal8Bit("提示"), QString::fromLocal8Bit("文件保存成功"), QString::fromLocal8Bit("确定"));
    ui->pushButton_save->setEnabled(false);
    memset(fileSrc,0,FILE_SIZE_MAX);
}

//![CRC-16数组]
quint8  crc_ta_ArrayHi[]={
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0,
    0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
    0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0,
    0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40,
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1,
    0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41,
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1,
    0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0,
    0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40,
    0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1,
    0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40,
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0,
    0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40,
    0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0,
    0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40,
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0,
    0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
    0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0,
    0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
    0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0,
    0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40,
    0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1,
    0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0,
    0x80, 0x41, 0x00, 0xC1, 0x81, 0x40};

quint8  crc_ta_ArrayLo[]={
    0x00, 0xC0, 0xC1, 0x01, 0xC3, 0x03, 0x02, 0xC2, 0xC6, 0x06,
    0x07, 0xC7, 0x05, 0xC5, 0xC4, 0x04, 0xCC, 0x0C, 0x0D, 0xCD,
    0x0F, 0xCF, 0xCE, 0x0E, 0x0A, 0xCA, 0xCB, 0x0B, 0xC9, 0x09,
    0x08, 0xC8, 0xD8, 0x18, 0x19, 0xD9, 0x1B, 0xDB, 0xDA, 0x1A,
    0x1E, 0xDE, 0xDF, 0x1F, 0xDD, 0x1D, 0x1C, 0xDC, 0x14, 0xD4,
    0xD5, 0x15, 0xD7, 0x17, 0x16, 0xD6, 0xD2, 0x12, 0x13, 0xD3,
    0x11, 0xD1, 0xD0, 0x10, 0xF0, 0x30, 0x31, 0xF1, 0x33, 0xF3,
    0xF2, 0x32, 0x36, 0xF6, 0xF7, 0x37, 0xF5, 0x35, 0x34, 0xF4,
    0x3C, 0xFC, 0xFD, 0x3D, 0xFF, 0x3F, 0x3E, 0xFE, 0xFA, 0x3A,
    0x3B, 0xFB, 0x39, 0xF9, 0xF8, 0x38, 0x28, 0xE8, 0xE9, 0x29,
    0xEB, 0x2B, 0x2A, 0xEA, 0xEE, 0x2E, 0x2F, 0xEF, 0x2D, 0xED,
    0xEC, 0x2C, 0xE4, 0x24, 0x25, 0xE5, 0x27, 0xE7, 0xE6, 0x26,
    0x22, 0xE2, 0xE3, 0x23, 0xE1, 0x21, 0x20, 0xE0, 0xA0, 0x60,
    0x61, 0xA1, 0x63, 0xA3, 0xA2, 0x62, 0x66, 0xA6, 0xA7, 0x67,
    0xA5, 0x65, 0x64, 0xA4, 0x6C, 0xAC, 0xAD, 0x6D, 0xAF, 0x6F,
    0x6E, 0xAE, 0xAA, 0x6A, 0x6B, 0xAB, 0x69, 0xA9, 0xA8, 0x68,
    0x78, 0xB8, 0xB9, 0x79, 0xBB, 0x7B, 0x7A, 0xBA, 0xBE, 0x7E,
    0x7F, 0xBF, 0x7D, 0xBD, 0xBC, 0x7C, 0xB4, 0x74, 0x75, 0xB5,
    0x77, 0xB7, 0xB6, 0x76, 0x72, 0xB2, 0xB3, 0x73, 0xB1, 0x71,
    0x70, 0xB0, 0x50, 0x90, 0x91, 0x51, 0x93, 0x53, 0x52, 0x92,
    0x96, 0x56, 0x57, 0x97, 0x55, 0x95, 0x94, 0x54, 0x9C, 0x5C,
    0x5D, 0x9D, 0x5F, 0x9F, 0x9E, 0x5E, 0x5A, 0x9A, 0x9B, 0x5B,
    0x99, 0x59, 0x58, 0x98, 0x88, 0x48, 0x49, 0x89, 0x4B, 0x8B,
    0x8A, 0x4A, 0x4E, 0x8E, 0x8F, 0x4F, 0x8D, 0x4D, 0x4C, 0x8C,
    0x44, 0x84, 0x85, 0x45, 0x87, 0x47, 0x46, 0x86, 0x82, 0x42,
    0x43, 0x83, 0x41, 0x81, 0x80, 0x40};
//![CRC-16计算]
quint16 MainWindow:: GetCRC16(char *puchMsg,unsigned int usDataLen)
{
    quint8 uchCRCHi = 0xFF;  /* CRC  的高字节初始化  */
    quint8 uchCRCLo = 0xFF;  /* CRC  的低字节初始化  */
    quint8 uIndex ;  	   /* CRC 查询表索引       */
    quint16 u16Result;
    quint16 i = 0;
    while(usDataLen--)  			 /* 完成整个报文缓冲区   */
    {
        uIndex = uchCRCLo ^ (quint8)puchMsg[i] ;  /*  计算  CRC      */
        i++;
        uchCRCLo = uchCRCHi ^ crc_ta_ArrayHi[uIndex];
        uchCRCHi = crc_ta_ArrayLo[uIndex];
    }
    u16Result = uchCRCHi;
    u16Result <<= 8;
    u16Result += uchCRCLo;
    return u16Result;
}

//![SUM-16计算]
quint16 MainWindow:: GetSUM16(char *puchMsg,unsigned int usDataLen)
{
    quint16 sumCheck = 0;
    unsigned int i = 0;
    for(i = 0;i<usDataLen;i++)
    {
        sumCheck += (quint8)(*(puchMsg + i));
    }
    return sumCheck;
}
